catkit/hardware/mcpherson/stuf_monochromator.py

```diff
@@ -0,0 +1,157 @@
+# TODO: This doesn't belong in catkit.
+
+import enum
+import time
+import warnings
+
+from catkit.interfaces.Instrument import Instrument
+from catkit.hardware.mcpherson.device_driver_747 import McPherson747
+from catkit.hardware.mcpherson.scan_controller_789a4 import McPherson789A4WithLimitSwitches
+from catkit.hardware.pyvisa_instrument import DEFAULT_POLL_TIMEOUT
+
+
+# STUF model: 234/302
+# STUF gratings: 355-107853-1 1200 g/mm concave corrected grating, Al+MgF2 coated, master holographic
+#                355-110987-1 600 g/mm concave corrected grating, Al+MgF2 coated, master holographic
+
+# For 234/302:
+# 600 g/mm => 4nm/rev
+# 1200 g/mm => 2nm/rev
+
+
+class Grating(enum.Enum):
+    def __init__(self, g_per_mm, nm_rev, position):
+        self.g_per_mm = g_per_mm
+        self.nm_rev = nm_rev
+        self.position = position
+
+    @classmethod
+    def _missing_(cls, value):
+        for item in cls:
+            if value in (item.g_per_mm, item.position):  # Luckily these are unique.
+                return item
+
+    # TODO: Check correctness of positions.
+    A = (6000, 4, 1)
+    B = (1200, 2, 2)
+
+
+class StufMonochromator(Instrument):
+
+    GRATING_DEVICE_NUMBER = 3
+
+    def initialize(self, gratings, kwargs_747, kwargs_789a4):
+        self.gratings = gratings
+
+        self.grating_controller = McPherson747(kwargs_747)
+        self.scan_controller = McPherson789A4WithLimitSwitches(kwargs_789a4)
+
+        self._current_grating = None
+
+    def manage_controllers(self):
+        with self.grating_controller:
+            with self.scan_controller:
+                yield
+
+    def _open(self):
+        self.manage_controllers()
+        return self
+
+    def _close(self):
+        try:
+            self.manage_controllers()
+        finally:
+            self._current_grating = None
+
+    @property
```

Comment:
@mgennaro When you get a mo can you take a scan over the functionality, please? Reading the function names & docstrings should be enough. I'm just looking to see if the functionality is as expected/desired or whether there's stuf(f) missing? Cheers